File: src/main/scala/biosimilarity/slmc/util/Namegen.scala
package biosimilarity.slmc.util

object Namegen {
  val bnInit = "$name"
  val fnInit = "#name"
  var count = 0

  // Determines if a string corresponds to a bound name
  def isBoundAlias(s: String): Boolean = {
    s.substring(0, Math.min(s.length, bnInit.length)).equals(bnInit)
  }

  // Returns a new bound name
  def generateBoundName(): String = {
    val res = bnInit + count
    count += 1
    res
  }

  // Generates a fresh name
  def freshAlias() = {
    val res = fnInit + count
    count += 1
    res
  }
  
  def freshRestrictions(size: Int) = {
    new Array[Any](size).map((_) => generateBoundName())
  }

  def freshAliases(size: Integer): List[String] = {
    for(i <- Range(0, size).toList) yield freshAlias()
  }
}

File: src/main/ocaml/namegen.ml
(*** Module that handles the generation of bound and fresh names ***)

(*** Constant declaration ***)

let bn_init = "$name"
let fn_init = "#name"
let count = ref 0

(***)

(*** Determines if a string corresponds to a bound name ***)

let is_bname s = (String.sub s 0 (min (String.length s) (String.length bn_init))) = bn_init

(*** Returns a new bound name ***)

let gen_bname () = 
  let res = bn_init^(string_of_int !count) in
  incr count; 
  res

(*** Generates a fresh name ***)

let fresh_name () =
  let res = fn_init^(string_of_int !count) in
  incr count;
  res

(*** Generates a list of fresh names ***)

let rec fresh_names size i =
  if i = size then
    []
  else
    (fresh_name())::(fresh_names size (i+1))

(***)
